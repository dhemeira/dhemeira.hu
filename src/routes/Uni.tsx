import { useEffect, useState } from 'react';
import { cookieValue } from '../utils';
import { AcademicCalendar, TypeOfWeek } from '../utils/academicCalendar';
import { Card } from '../components/Card';
import { clsx } from 'clsx';

export const Uni = () => {
  const [dates, setDates] = useState(new AcademicCalendar());
  const [type, setType] = useState(TypeOfWeek.Break);
  const formattedDate = new Date().toLocaleDateString('hu-Hu');

  useEffect(() => {
    const setValues = (dates: AcademicCalendar) => {
      setDates(dates);
      setType(AcademicCalendar.typeOfWeek(formattedDate, dates));
    };

    const fetchData = async () => {
      await fetch('/api/date', {
        method: 'GET',
      })
        .then((response) => {
          return response.json();
        })
        .then((data: AcademicCalendar) => {
          setValues(data);

          const expireDate = new Date(
            formattedDate < data.semester_end_date ? data.semester_end_date : data.exam_end_date
          );
          expireDate.setDate(expireDate.getDate() + 1);
          expireDate.setHours(0, 0, 1);

          document.cookie = `dates=${JSON.stringify(
            data
          )};expires=${expireDate.toUTCString()};sameSite=Lax`;
        })
        .catch(() => {
          console.error('Date fetch failed');
        });
    };

    const cookie = cookieValue('dates');
    if (cookie) {
      setValues(JSON.parse(cookie));
    } else fetchData();
  }, [formattedDate]);

  return (
    <div
      className={clsx(
        'container',
        'flex',
        'flex-col',
        'gap-8',
        'md:gap-16',
        'justify-center',
        'items-center'
      )}>
      <h1 className="text-4xl sm:text-5xl font-bold tracking-tighter">University</h1>
      <div className="flex">
        <Card>
          <>
            <p className="text-sm">{formattedDate}</p>
            <div className="flex flex-col justify-center items-center pb-4 text-light-text dark:text-dark-text">
              <p className="text-5xl">{AcademicCalendar.weekEmoji(type)}</p>
              <p className="text-4xl font-semibold leading-tight">
                {AcademicCalendar.weekTitle(formattedDate, dates, type)}
              </p>
              <p className="">{AcademicCalendar.weekType(type)}</p>
            </div>

            <div
              className={clsx(
                'flex flex-col',
                'justify-center',
                'items-center',
                'gap-4',
                'border-t',
                'border-t-light-text/20',
                'dark:border-t-dark-text/20',
                'pt-4'
              )}>
              <p className="inline-flex flex-col justify-center items-center">
                <span className="text-sm">Study period</span>
                <span>
                  {dates.semester_start_date} - {dates.semester_end_date}
                </span>
              </p>
              <p className="inline-flex flex-col justify-center items-center gap-x-1">
                <span className="text-sm">Exam period</span>
                <span>
                  {dates.exam_start_date} - {dates.exam_end_date}
                </span>
              </p>
            </div>
          </>
        </Card>
      </div>
    </div>
  );
};
