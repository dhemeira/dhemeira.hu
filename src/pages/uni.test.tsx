import React from 'react';
import { cleanup, render, waitFor, screen } from '@testing-library/react';
import Uni from './uni';

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () =>
      Promise.resolve({
        semester_start_date: '2024. 09. 09.',
        semester_end_date: '2024. 12. 15.',
        exam_start_date: '2024. 12. 16.',
        exam_end_date: '2025. 02. 08.',
      }),
  })
) as jest.Mock;

describe('University', () => {
  beforeEach(() => {
    const mockDate = new Date('2024-09-08');
    jest.useFakeTimers();
    jest.setSystemTime(mockDate);
  });
  afterEach(() => {
    cleanup();
  });
  afterAll(() => {
    jest.useRealTimers();
  });

  test('date is mocked', async () => {
    const date = new Date();
    expect(date.toISOString()).toBe('2024-09-08T00:00:00.000Z');
  });

  test('renders correctly', async () => {
    const { container } = render(<Uni />);

    // Wait for the loading text to disappear
    await waitFor(() => expect(screen.queryByText('Loading')).not.toBeInTheDocument());

    expect(container.firstChild).toMatchSnapshot();
  });

  //Emoji     Title      Type
  //‚è≥        Loading    Loading...
  //üìñ        Week 1     of study period
  //üìù        Week 1     of exam period
  //üèñÔ∏è        Break      Enjoy!
  test('week is study period', async () => {
    const mockDate = new Date('2024-09-09');
    jest.setSystemTime(mockDate);

    const { getByTestId } = render(<Uni key="asd" />);

    // Wait for the loading text to disappear
    await waitFor(() => expect(screen.queryByText('Loading')).not.toBeInTheDocument());

    expect(getByTestId('weekEmoji')).toHaveTextContent('üìñ');
    expect(getByTestId('weekTitle')).toHaveTextContent('Week 1');
    expect(getByTestId('weekType')).toHaveTextContent('of study period');
  });
  test('week is exam period', async () => {
    const mockDate = new Date('2024-12-16');
    jest.setSystemTime(mockDate);

    const { getByTestId } = render(<Uni key="asd" />);

    // Wait for the loading text to disappear
    await waitFor(() => expect(screen.queryByText('Loading')).not.toBeInTheDocument());

    expect(getByTestId('weekEmoji')).toHaveTextContent('üìù');
    expect(getByTestId('weekTitle')).toHaveTextContent('Week 1');
    expect(getByTestId('weekType')).toHaveTextContent('of exam period');
  });
  test('week is break', async () => {
    const mockDate = new Date('2025-02-09');
    jest.setSystemTime(mockDate);

    const { getByTestId } = render(<Uni key="asd" />);

    // Wait for the loading text to disappear
    await waitFor(() => expect(screen.queryByText('Loading')).not.toBeInTheDocument());

    expect(getByTestId('weekEmoji')).toHaveTextContent('üèñÔ∏è');
    expect(getByTestId('weekTitle')).toHaveTextContent('Break');
    expect(getByTestId('weekType')).toHaveTextContent('Enjoy!');
  });
});
