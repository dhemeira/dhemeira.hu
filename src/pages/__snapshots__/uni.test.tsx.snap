// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`University Page card renders correctly 1`] = `
<div
  class="bg-light-primary/5 dark:bg-dark-primary/5 from-30% bg-gradient-to-tl from-light-secondary/30 dark:from-dark-secondary/5 to-light-secondary/5 dark:to-dark-secondary/30 rounded-xl shadow border border-light-text/15 dark:border-dark-text/15 text-light-text/80 dark:text-dark-text/80 tabular-nums overflow-hidden relative text-center flex flex-col justify-between items-center gap-2 px-10 md:px-14 pt-4 pb-6"
  data-testid="uniCard"
>
  <span
    class="absolute inset-x-0 bottom-0 h-2 bg-gradient-to-r to-light-primary from-light-secondary dark:to-dark-primary dark:from-dark-secondary"
  />
  <span
    class="text-sm"
  >
    2024. 09. 08.
  </span>
  <div
    class="flex flex-col justify-center items-center gap-1 pb-4 text-light-text dark:text-dark-text"
  >
    <span
      class="text-5xl"
      data-testid="weekEmoji"
    >
      üèñÔ∏è
    </span>
    <span
      class="text-4xl font-semibold leading-none"
      data-testid="weekTitle"
    >
      Break
    </span>
    <span
      data-testid="weekType"
    >
      Enjoy!
    </span>
  </div>
  <div
    class="flex flex-col justify-center items-center gap-4 border-t border-t-light-text/20 dark:border-t-dark-text/20 pt-4"
  >
    <span
      class="inline-flex flex-col justify-center items-center"
    >
      <span
        class="text-sm"
      >
        Study period
      </span>
      <span
        class="inline-flex flex-row gap-1 justify-center items-center"
      >
        2024. 09. 09.
        <span>
          -
        </span>
        2024. 12. 15.
      </span>
    </span>
    <span
      class="inline-flex flex-col justify-center items-center gap-x-1"
    >
      <span
        class="text-sm"
      >
        Exam period
      </span>
      <span
        class="inline-flex flex-row gap-1 justify-center items-center"
      >
        2024. 12. 16.
        <span>
          -
        </span>
        2025. 02. 08.
      </span>
    </span>
  </div>
</div>
`;
